src/util/codec/mysql/json.rs

```diff
@@ -0,0 +1,472 @@
+            Json::Literal(l) => {
+                match l {
+                    JSON_LITERAL_NIL => serializer.serialize_none(),
+                    JSON_LITERAL_TRUE => serializer.serialize_bool(true),
+                    _ => serializer.serialize_bool(false),
+                }
+            }
+            Json::String(ref s) => serializer.serialize_str(s),
+            Json::Object(ref obj) => {
+                let mut map = try!(serializer.serialize_map(Some(obj.len())));
+                for (k, v) in obj {
+                    try!(map.serialize_entry(k, v));
+                }
+                map.end()
+            }
+            Json::Array(ref array) => {
+                let mut tup = try!(serializer.serialize_tuple(array.len()));
+                for item in array {
+                    try!(tup.serialize_element(item));
+                }
+                tup.end()
+            }
+            Json::Double(d) => serializer.serialize_f64(d),
+            Json::I64(d) => serializer.serialize_i64(d),
+        }
+    }
+}
+
+fn encode_literal(data: u8, buf: &mut Vec<u8>) {
+    buf.push(data);
+}
+
+fn decode_literal(data: &[u8]) -> Result<Json> {
+    Ok(Json::Literal(data[0]))
+}
+
+fn encode_str(data: String, buf: &mut Vec<u8>) {
+    buf.encode_var_u64(data.len() as u64).unwrap();
+    buf.append(&mut data.into_bytes());
+}
+
+fn decode_str(mut data: &[u8]) -> Result<Json> {
+    let length = try!(data.decode_var_u64());
+    let encode_value = &data[0..(length as usize)];
+    match str::from_utf8(encode_value) {
+        Ok(v) => {
+            let value = String::from(v);
+            Ok(Json::String(value))
+        }
+        Err(e) => Err(invalid_type!("Invalid UTF-8 sequence:{}", e)),
+    }
+}
+
+fn encode_double(data: f64, buf: &mut Vec<u8>) {
+    buf.encode_f64_with_little_endian(data).unwrap();
+}
+
+fn decode_double(mut data: &[u8]) -> Result<Json> {
+    let value = try!(data.decode_f64_with_little_endian());
+    Ok(Json::Double(value))
+}
+
+fn encode_i64(data: i64, buf: &mut Vec<u8>) {
+    buf.encode_i64_with_little_endian(data).unwrap();
+}
+
+fn decode_i64(mut data: &[u8]) -> Result<Json> {
+    let value = try!(data.decode_i64_with_little_endian());
+    Ok(Json::I64(value))
+}
+
+fn encode_obj(data: BTreeMap<String, Json>, buf: &mut Vec<u8>) {
+    // object: element-count size key-entry* value-entry* key* value*
+    // element-count ::= uint32 number of members in object or number of elements in array
+    let element_count = data.len();
+    let element_count_len = LENGTH_U32;
+    // size ::= uint32 number of bytes in the binary representation of the object or array
+    let size_len = LENGTH_U32;
+    // key-entry ::= key-offset(uint32) key-length(uint16)
+    let key_entries_len = LENGTH_KEY_ENTRY * element_count;
+    // value-entry ::= type(byte) offset-or-inlined-value(uint32)
+    let value_entries_len = LENGTH_VALUE_ENTRY * element_count;
+    let mut keys = Vec::with_capacity(element_count);
```

Comment:
Since we couldn't independently take ownership of the value  from a map. It seems not a good idea since I would like to `take`  the `key,value`  from the `data` instead of `clone`.  @hicqu